File: src/chapters/numerical-calculus/content/code/simpson.py
import numpy as np


def simpson(f, a, b, n=100):
    """Composite Simpson's rule (n forced even) on [a, b]."""
    if n % 2:
        n += 1
    x = np.linspace(a, b, n + 1)
    y = f(x)
    h = (b - a) / n
    return h / 3 * (y[0] + y[-1] + 4 * y[1:-1:2].sum() + 2 * y[2:-1:2].sum())


if __name__ == "__main__":
    import math
    print("int_0^1 e^x dx ~", simpson(np.exp, 0, 1, 100), " exact =", math.e - 1)

File: src/chapters/numerical-calculus/content/code/simpson.jl
function simpson(f, a, b, n = 100)
    isodd(n) && (n += 1)
    h = (b - a) / n
    s = f(a) + f(b)
    for i in 1:n-1
        s += (isodd(i) ? 4 : 2) * f(a + i*h)
    end
    return h / 3 * s
end

println("int_0^1 e^x dx = ", simpson(exp, 0.0, 1.0, 100))
